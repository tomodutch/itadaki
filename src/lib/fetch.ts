export async function fetchWithTimeout(
    url: string,
    options: RequestInit = {},
    timeoutMs = 5000
): Promise<Response> {
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), timeoutMs);
    // If the caller passed a signal, combine it with our timeout signal
    const signals = [controller.signal];
    if (options.signal) signals.push(options.signal);

    const finalController = new AbortController();
    signals.forEach((s) =>
        s.addEventListener("abort", () => finalController.abort())
    );

    try {
        const response = await fetch(url, {
            ...options,
            signal: finalController.signal,
        });
        clearTimeout(timeout);
        return response;
    } catch (err) {
        clearTimeout(timeout);
        throw err;
    }
}