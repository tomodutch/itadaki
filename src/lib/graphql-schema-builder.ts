import SchemaBuilder from '@pothos/core';
import prisma from '@/lib/prisma';
import PrismaPlugin from '@pothos/plugin-prisma';
import { PrismaObjectFieldBuilder } from '@pothos/plugin-prisma';
// This is the default location for the generator, but this can be
// customized as described above.
// Using a type only import will help avoid issues with undeclared
// exports in esm mode
import type PrismaTypes from '@/db/generated/pothos';

const builder = new SchemaBuilder<{
  PrismaTypes: PrismaTypes;
}>({
  plugins: [PrismaPlugin],
  prisma: {
    client: prisma,
    // defaults to false, uses /// comments from prisma schema as descriptions
    // for object types, relations and exposed fields.
    // descriptions can be omitted by setting description to false
    // exposeDescriptions: false | { models: boolean, fields: boolean },
    // use where clause from prismaRelatedConnection for totalCount (defaults to true)
    filterConnectionTotalCount: true,
    // warn when not using a query parameter correctly
    onUnusedQuery: process.env.NODE_ENV === 'production' ? null : 'warn',
  },
});

builder.prismaObject('User', {
  fields: (t: PrismaObjectFieldBuilder<PrismaTypes, "User">) => ({
    id: t.exposeID('id'),
    name: t.exposeString('name', { nullable: true }),
    image: t.exposeString('image', { nullable: true }),
  }),
});

builder.queryField('userById', (t) =>
  t.prismaField({
    type: 'User',
    args: {
      id: t.arg.string({ required: true }),
    },
    resolve: (query, _root, args) =>
      prisma.user.findUnique({
        ...query,
        where: { id: args.id },
      }),
  }),
);

builder.mutationField('createUser', (t) =>
  t.prismaField({
    type: 'User',
    args: {
      email: t.arg.string({ required: true }),
      name: t.arg.string({ required: false }),
      image: t.arg.string({ required: false }),
    },
    resolve: async (query, _root, args) => {
      return prisma.user.create({
        ...query,
        data: {
          email: args.email,
          name: args.name ?? null,
          image: args.image ?? null,
        },
      });
    },
  })
);

// Add explicit root Query type declaration
builder.queryType({});
builder.mutationType({});

export default builder;