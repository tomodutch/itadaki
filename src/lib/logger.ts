import winston from "winston";

const logger = winston.createLogger({
    level: 'info',
    format: winston.format.combine(
        winston.format.timestamp(),
        winston.format.errors({ stack: true }),
        winston.format.json()
    ),
    defaultMeta: { service: 'itadaki-web' },
    transports: [
        new winston.transports.Console({
            format: winston.format.combine(
                winston.format.colorize(),
                winston.format.printf(({ level, message, timestamp, ...meta }) => {
                    return `${timestamp} [${level}]: ${message} ${Object.keys(meta).length ? JSON.stringify(meta) : ""
                        }`;
                })
            )
        })
    ]
});

export default logger;