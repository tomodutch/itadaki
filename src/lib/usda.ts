import { FoodTemplate } from "@/db/generated/prisma";

export type USDAFoodTemplate = Omit<FoodTemplate, "id" | "userId" | "brand" | "createdAt" | "deletedAt" | "updatedAt"> & { origin: "usda" }

export function mapResponseToFoodTemplates(data: FoodSearchResponse): USDAFoodTemplate[] {
    const foodTemplates: USDAFoodTemplate[] = [];
    for (const food of data.foods) {
        const foodTemplate: USDAFoodTemplate = {
            name: food.commonNames || food.description,
            servingSize: 100,
            servingUnit: "g",
            barCode: null,
            calories: 0,
            protein: null,
            carbs: null,
            sugar: null,
            fiber: null,
            fat: null,
            saturatedFat: null,
            transFat: null,
            cholesterol: null,
            sodium: null,
            potassium: null,
            vitaminA: null,
            vitaminC: null,
            vitaminD: null,
            vitaminE: null,
            vitaminK: null,
            vitaminB1: null,
            vitaminB2: null,
            vitaminB3: null,
            vitaminB5: null,
            vitaminB6: null,
            vitaminB7: null,
            vitaminB9: null,
            vitaminB12: null,
            calcium: null,
            iron: null,
            magnesium: null,
            phosphorus: null,
            zinc: null,
            copper: null,
            manganese: null,
            selenium: null,
            alcohol: null,
            caffeine: null,
            origin: "usda",
            originId: food.fdcId.toString()
        };

        for (const nutrient of food.foodNutrients) {
            const field = nutrientIdToFieldMap[nutrient.nutrientId];
            if (field && typeof nutrient.value === "number") {
                foodTemplate[field] = nutrient.value;
            }

            if ([2047, 1008].indexOf(nutrient.nutrientId) > -1 && typeof nutrient.value === "number") {
                foodTemplate.calories = nutrient.value;
            }
        }

        foodTemplates.push(foodTemplate);
    }
    return foodTemplates;
}

export interface FoodSearchResponse {
    totalHits: number;
    currentPage: number;
    totalPages: number;
    pageList: number[];
    foodSearchCriteria: FoodSearchCriteria;
    foods: Food[];
}

export interface FoodSearchCriteria {
    dataType: string[];       // e.g. ["Foundation"]
    query: string;            // e.g. "lentil"
    generalSearchInput: string;
    pageNumber: number;
    sortBy: string;           // e.g. "dataType.keyword"
    sortOrder: string;        // e.g. "asc"
    numberOfResultsPerPage: number;
    pageSize: number;
    requireAllWords: boolean;
    foodTypes: string[];      // e.g. ["Foundation"]
}

export interface Food {
    fdcId: number;
    description: string;
    commonNames: string;
    additionalDescriptions: string;
    dataType: string;
    ndbNumber: number;
    publishedDate: string;            // ISO date string
    foodCategory: string;
    mostRecentAcquisitionDate: string; // ISO date string
    allHighlightFields: string;
    score: number;
    microbes: unknown[];
    foodNutrients: FoodNutrient[];
}

export interface FoodNutrient {
    nutrientId: number;
    nutrientName: string;
    nutrientNumber: string;
    unitName: string;
    derivationCode: string;
    derivationDescription: string;
    derivationId: number;
    value: number;
    foodNutrientSourceId: number;
    foodNutrientSourceCode: string;
    foodNutrientSourceDescription: string;
    rank: number;
    indentLevel: number;
    foodNutrientId: number;
    dataPoints: number;
    min: number;
    max: number;
    median: number;
}

type NumericFoodTemplateFields = {
    [K in keyof FoodTemplate]: FoodTemplate[K] extends number | null ? K : never;
}[keyof FoodTemplate];

const nutrientIdToFieldMap: Record<number, NumericFoodTemplateFields> = {
    1003: "protein",
    1004: "fat",
    1005: "carbs",
    2000: "sugar",
    1079: "fiber",
    1258: "saturatedFat",
    1257: "transFat",
    1253: "cholesterol",
    1093: "sodium",
    1092: "potassium",
    1106: "vitaminA",
    1162: "vitaminC",
    1114: "vitaminD",
    1109: "vitaminE",
    1185: "vitaminK",
    1165: "vitaminB1",
    1166: "vitaminB2",
    1167: "vitaminB3",
    1170: "vitaminB5",
    1175: "vitaminB6",
    1176: "vitaminB7",
    1177: "vitaminB9",
    1178: "vitaminB12",
    1087: "calcium",
    1089: "iron",
    1090: "magnesium",
    1091: "phosphorus",
    1095: "zinc",
    1098: "copper",
    1101: "manganese",
    1103: "selenium",
    1100: "alcohol",
    1057: "caffeine"
};
