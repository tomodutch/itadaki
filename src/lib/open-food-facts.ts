
import { FoodTemplate } from "@/db/generated/prisma";
import { OpenFoodFactsApiResponse } from "@/lib/types/open-food-facts";
import { fetchWithTimeout } from "./fetch";

export async function getByCode(code: string) {
        const url = `https://world.openfoodfacts.org/api/v2/product/${code}`;
    
    const options: RequestInit = {
        method: "GET",
        headers: {
            "Accept": "application/json",
            "User-Agent": "itadaki.app - itadaki@tomodutch.com"
        }
    };

    const resp = await fetchWithTimeout(url, options);
    if (resp.status !== 200) {
        return;
    }

    return mapOFFToFoodTemplate(await resp.json());
}

export type OFFFoodTemplate = Omit<FoodTemplate, "id" | "userId" | "brand" | "createdAt" | "deletedAt" | "updatedAt"> & { origin: "open-food-facts" }
export function mapOFFToFoodTemplate(data: OpenFoodFactsApiResponse): OFFFoodTemplate {
    const { product } = data;
    const n = product.nutriments;

    const foodTemplate: OFFFoodTemplate = {
        name: product.product_name,
        servingSize: typeof product.serving_quantity === "number" ? product.serving_quantity : Number.parseFloat(product.serving_quantity),
        servingUnit: product.serving_quantity_unit,
        barCode: product.code,
        calories: n["energy-kcal_serving"] ?? null,
        protein: n.proteins_serving ?? null,
        carbs: n.carbohydrates_serving ?? null,
        sugar: n.sugars_serving ?? null,
        fiber: n.fiber_serving ?? null,
        fat: n.fat_serving ?? null,
        saturatedFat: n["saturated-fat_serving"] ?? null,
        transFat: null, // Not present in `Nutriments`
        cholesterol: null, // Not present in `Nutriments`
        sodium: n.sodium_serving ?? null,
        potassium: null, // Not present in `Nutriments`
        vitaminA: null,
        vitaminC: null,
        vitaminD: null,
        vitaminE: null,
        vitaminK: null,
        vitaminB1: null,
        vitaminB2: null,
        vitaminB3: null,
        vitaminB5: null,
        vitaminB6: null,
        vitaminB7: null,
        vitaminB9: null,
        vitaminB12: null,
        calcium: null,
        iron: null,
        magnesium: null,
        phosphorus: null,
        zinc: null,
        copper: null,
        manganese: null,
        selenium: null,
        alcohol: null,
        caffeine: null,
        origin: "open-food-facts",
        originId: product.id
    };

    return foodTemplate;
}