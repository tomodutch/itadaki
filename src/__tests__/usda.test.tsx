import { expect, test } from "vitest";
import response from "./fixtures/usda_resp.json";
import cabbageResponse from "./fixtures/usda_cabbage.json";
import baconResponse from "./fixtures/usda_bacon.json";

import { FoodSearchResponse, mapResponseToFoodTemplates } from "@/lib/usda";

test("mapping response", () => {
    const foodTemplates = mapResponseToFoodTemplates(response as FoodSearchResponse);
    expect(foodTemplates).toMatchSnapshot();
});

test("map cabbage response", () => {
    const foodTemplates = mapResponseToFoodTemplates(cabbageResponse as FoodSearchResponse);
    expect(foodTemplates).toMatchSnapshot();
});

test("map bacon response", () => {
    const foodTemplates = mapResponseToFoodTemplates(baconResponse as FoodSearchResponse);
    expect(foodTemplates).toMatchSnapshot();
});