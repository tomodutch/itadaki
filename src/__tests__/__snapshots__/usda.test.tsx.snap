// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`map bacon response 1`] = `
[
  {
    "alcohol": null,
    "barCode": null,
    "caffeine": null,
    "calcium": 13,
    "calories": 500,
    "carbs": 2.1,
    "cholesterol": null,
    "copper": 0.13,
    "fat": 36.5,
    "fiber": null,
    "iron": 1.28,
    "magnesium": 36.5,
    "manganese": 0.02,
    "name": "Pork, cured, bacon, cooked, restaurant",
    "origin": "usda",
    "originId": "749420",
    "phosphorus": 415,
    "potassium": 557,
    "protein": 40.9,
    "saturatedFat": 12.6,
    "selenium": null,
    "servingSize": 100,
    "servingUnit": "g",
    "sodium": 1830,
    "sugar": null,
    "transFat": 0.15,
    "vitaminA": null,
    "vitaminB1": null,
    "vitaminB12": null,
    "vitaminB2": null,
    "vitaminB3": null,
    "vitaminB5": null,
    "vitaminB6": null,
    "vitaminB7": null,
    "vitaminB9": null,
    "vitaminC": null,
    "vitaminD": null,
    "vitaminE": null,
    "vitaminK": null,
    "zinc": 3.56,
  },
]
`;

exports[`map cabbage response 1`] = `
[
  {
    "alcohol": null,
    "barCode": null,
    "caffeine": null,
    "calcium": 41.8,
    "calories": 31.4,
    "carbs": 6.38,
    "cholesterol": null,
    "copper": 0,
    "fat": 0.228,
    "fiber": null,
    "iron": 0.0662,
    "magnesium": 13.9,
    "manganese": 0.248,
    "name": "Cabbage, green, raw",
    "origin": "usda",
    "originId": "2346407",
    "phosphorus": 26.9,
    "potassium": 207,
    "protein": 0.961,
    "saturatedFat": null,
    "selenium": null,
    "servingSize": 100,
    "servingUnit": "g",
    "sodium": 16.1,
    "sugar": null,
    "transFat": null,
    "vitaminA": null,
    "vitaminB1": null,
    "vitaminB12": null,
    "vitaminB2": null,
    "vitaminB3": null,
    "vitaminB5": null,
    "vitaminB6": 0.138,
    "vitaminB7": null,
    "vitaminB9": null,
    "vitaminC": 40.3,
    "vitaminD": null,
    "vitaminE": null,
    "vitaminK": 59.4,
    "zinc": 0.211,
  },
  {
    "alcohol": null,
    "barCode": null,
    "caffeine": null,
    "calcium": 31,
    "calories": 34.1,
    "carbs": 6.79,
    "cholesterol": null,
    "copper": 0.0035,
    "fat": 0.214,
    "fiber": null,
    "iron": 0,
    "magnesium": 13.3,
    "manganese": 0.309,
    "name": "Cabbage, red, raw",
    "origin": "usda",
    "originId": "2346408",
    "phosphorus": 35.8,
    "potassium": 269,
    "protein": 1.24,
    "saturatedFat": null,
    "selenium": null,
    "servingSize": 100,
    "servingUnit": "g",
    "sodium": 11.7,
    "sugar": null,
    "transFat": null,
    "vitaminA": null,
    "vitaminB1": null,
    "vitaminB12": null,
    "vitaminB2": null,
    "vitaminB3": null,
    "vitaminB5": null,
    "vitaminB6": 0.179,
    "vitaminB7": null,
    "vitaminB9": null,
    "vitaminC": 53.9,
    "vitaminD": null,
    "vitaminE": null,
    "vitaminK": 32.3,
    "zinc": 0.245,
  },
  {
    "alcohol": null,
    "barCode": null,
    "caffeine": null,
    "calcium": 61.9,
    "calories": 20.3,
    "carbs": 3.51,
    "cholesterol": null,
    "copper": 0.0276,
    "fat": 0.234,
    "fiber": 1.26,
    "iron": 0.439,
    "magnesium": 13.6,
    "manganese": 0.196,
    "name": "Cabbage, bok choy, raw",
    "origin": "usda",
    "originId": "2685572",
    "phosphorus": 30.5,
    "potassium": 228,
    "protein": 1.02,
    "saturatedFat": null,
    "selenium": null,
    "servingSize": 100,
    "servingUnit": "g",
    "sodium": 14.3,
    "sugar": null,
    "transFat": null,
    "vitaminA": null,
    "vitaminB1": null,
    "vitaminB12": null,
    "vitaminB2": null,
    "vitaminB3": null,
    "vitaminB5": null,
    "vitaminB6": null,
    "vitaminB7": null,
    "vitaminB9": 68,
    "vitaminC": 30.3,
    "vitaminD": null,
    "vitaminE": null,
    "vitaminK": null,
    "zinc": 0.274,
  },
  {
    "alcohol": null,
    "barCode": null,
    "caffeine": null,
    "calcium": 35,
    "calories": 0,
    "carbs": null,
    "cholesterol": null,
    "copper": 0.0238,
    "fat": null,
    "fiber": null,
    "iron": 0.255,
    "magnesium": 11.3,
    "manganese": 0.201,
    "name": "Cabbage, napa, leaf, destemmed, raw",
    "origin": "usda",
    "originId": "2727583",
    "phosphorus": 34.4,
    "potassium": 235,
    "protein": 1.06,
    "saturatedFat": null,
    "selenium": null,
    "servingSize": 100,
    "servingUnit": "g",
    "sodium": 12.6,
    "sugar": null,
    "transFat": null,
    "vitaminA": null,
    "vitaminB1": null,
    "vitaminB12": null,
    "vitaminB2": null,
    "vitaminB3": null,
    "vitaminB5": null,
    "vitaminB6": 0.0856,
    "vitaminB7": 2.72,
    "vitaminB9": null,
    "vitaminC": null,
    "vitaminD": null,
    "vitaminE": null,
    "vitaminK": null,
    "zinc": 0.253,
  },
]
`;

exports[`mapping response 1`] = `
[
  {
    "alcohol": null,
    "barCode": null,
    "caffeine": null,
    "calcium": 61.8,
    "calories": 360,
    "carbs": 62.2,
    "cholesterol": null,
    "copper": 0.838,
    "fat": 1.92,
    "fiber": null,
    "iron": 7.16,
    "magnesium": 107,
    "manganese": 1.57,
    "name": "Lentils, dry",
    "origin": "usda",
    "originId": "2644283",
    "phosphorus": 374,
    "potassium": 949,
    "protein": 23.6,
    "saturatedFat": null,
    "selenium": null,
    "servingSize": 100,
    "servingUnit": "g",
    "sodium": 0,
    "sugar": null,
    "transFat": null,
    "vitaminA": null,
    "vitaminB1": null,
    "vitaminB12": null,
    "vitaminB2": null,
    "vitaminB3": null,
    "vitaminB5": null,
    "vitaminB6": null,
    "vitaminB7": null,
    "vitaminB9": null,
    "vitaminC": null,
    "vitaminD": null,
    "vitaminE": null,
    "vitaminK": null,
    "zinc": 3.86,
  },
]
`;
